flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java

```diff
@@ -41,16 +42,49 @@
   private final FileScanTaskReader<T> fileScanTaskReader;
 
   private final InputFilesDecryptor inputFilesDecryptor;
-  private Iterator<FileScanTask> tasks;
+  private final CombinedScanTask combinedTask;
+
+  private Iterator<FileScanTask> fileTasksIterator;
   private CloseableIterator<T> currentIterator;
+  private int fileOffset;
+  private long recordOffset;
 
   public DataIterator(FileScanTaskReader<T> fileScanTaskReader, CombinedScanTask task,
                       FileIO io, EncryptionManager encryption) {
     this.fileScanTaskReader = fileScanTaskReader;
 
     this.inputFilesDecryptor = new InputFilesDecryptor(task, io, encryption);
-    this.tasks = task.files().iterator();
+    this.combinedTask = task;
+
+    this.fileTasksIterator = task.files().iterator();
     this.currentIterator = CloseableIterator.empty();
+
+    // fileOffset starts at -1 because we started
+    // from an empty iterator that is not from the split files.
+    this.fileOffset = -1;
+    this.recordOffset = 0L;
+  }
+
+  public void seek(int startingFileOffset, long startingRecordOffset) {
+    // skip files
+    Preconditions.checkArgument(startingFileOffset < combinedTask.files().size(),
+        "Checkpointed file offset is %d, while CombinedScanTask has %d files",
```

Comment:
For error messages, we state the problem up front and provide context. Usually that takes a form like "Invalid <something>: %s <additional help>" or "Cannot <action>: <reason and context>".

Here, this should be: `"Invalid offset for task %s with %d files: file=%d, record=%d", combinedTask, combinedTask.files().size(), startingFileOffset, startingRecordOffset`

Reply:
will update to `"Invalid startingFileOffset %d for CombinedScanTask with %d files: %s",
        startingFileOffset, combinedTask.files().size(), combinedTask`